File: dotnet/test/Zametek.Maths.Graphs.Primitives.Tests/Entities/NodeTests.cs
using Shouldly;
using System;
using Xunit;

namespace Zametek.Maths.Graphs.Tests
{
    public class NodeTests
    {
        [Fact]
        public void Node_GivenCtor_WithNullContent_ThenThrowsArgumentNullException()
        {
            Action act = () => new Node<int, Event<int>>(null);
            act.ShouldThrow<ArgumentNullException>();
        }

        [Fact]
        public void Node_GivenCtor_WithContentOnly_ThenNodeTypeIsNormal()
        {
            var node = new Node<int, Event<int>>(new Event<int>(1));

            node.NodeType.ShouldBe(NodeType.Normal);
            node.Id.ShouldBe(1);
        }

        [Fact]
        public void Node_GivenIncomingEdges_WhenStartNode_ThenThrowsInvalidOperationException()
        {
            var node = new Node<int, Event<int>>(NodeType.Start, new Event<int>(1));

            Action act = () => _ = node.IncomingEdges;
            act.ShouldThrow<InvalidOperationException>();
        }

        [Fact]
        public void Node_GivenIncomingEdges_WhenIsolatedNode_ThenThrowsInvalidOperationException()
        {
            var node = new Node<int, Event<int>>(NodeType.Isolated, new Event<int>(1));

            Action act = () => _ = node.IncomingEdges;
            act.ShouldThrow<InvalidOperationException>();
        }

        [Fact]
        public void Node_GivenOutgoingEdges_WhenEndNode_ThenThrowsInvalidOperationException()
        {
            var node = new Node<int, Event<int>>(NodeType.End, new Event<int>(1));

            Action act = () => _ = node.OutgoingEdges;
            act.ShouldThrow<InvalidOperationException>();
        }

        [Fact]
        public void Node_GivenOutgoingEdges_WhenIsolatedNode_ThenThrowsInvalidOperationException()
        {
            var node = new Node<int, Event<int>>(NodeType.Isolated, new Event<int>(1));

            Action act = () => _ = node.OutgoingEdges;
            act.ShouldThrow<InvalidOperationException>();
        }

        [Fact]
        public void Node_GivenSetNodeType_ThenNodeTypeChanges()
        {
            var node = new Node<int, Event<int>>(NodeType.Start, new Event<int>(1));

            node.SetNodeType(NodeType.Normal);

            node.NodeType.ShouldBe(NodeType.Normal);
        }

        [Fact]
        public void Node_GivenEquals_WhenSameIdTypeAndEdges_ThenEqual()
        {
            var node1 = new Node<int, Event<int>>(new Event<int>(1));
            node1.IncomingEdges.Add(10);
            node1.OutgoingEdges.Add(20);
            var node2 = new Node<int, Event<int>>(new Event<int>(1));
            node2.IncomingEdges.Add(10);
            node2.OutgoingEdges.Add(20);

            node1.Equals(node2).ShouldBeTrue();
            node1.GetHashCode().ShouldBe(node2.GetHashCode());
        }

        [Fact]
        public void Node_GivenEquals_WhenDifferentNodeType_ThenNotEqual()
        {
            var node1 = new Node<int, Event<int>>(NodeType.Normal, new Event<int>(1));
            var node2 = new Node<int, Event<int>>(NodeType.End, new Event<int>(1));

            node1.Equals(node2).ShouldBeFalse();
        }

        [Fact]
        public void Node_GivenEquals_WhenDifferentEdges_ThenNotEqual()
        {
            var node1 = new Node<int, Event<int>>(new Event<int>(1));
            node1.IncomingEdges.Add(10);
            var node2 = new Node<int, Event<int>>(new Event<int>(1));
            node2.IncomingEdges.Add(11);

            node1.Equals(node2).ShouldBeFalse();
        }

        [Fact]
        public void Node_GivenCloneObject_ThenTypeContentAndEdgesPreserved()
        {
            var content = new Event<int>(1, 2, 3);
            var node = new Node<int, Event<int>>(NodeType.Normal, content);
            node.IncomingEdges.Add(10);
            node.OutgoingEdges.Add(20);

            var clone = (Node<int, Event<int>>)node.CloneObject();

            clone.NodeType.ShouldBe(NodeType.Normal);
            clone.Content.ShouldNotBeSameAs(content);
            clone.Content.EarliestFinishTime.ShouldBe(2);
            clone.IncomingEdges.ShouldBe([10]);
            clone.OutgoingEdges.ShouldBe([20]);
        }

        [Fact]
        public void Node_GivenCloneObject_ThenEdgeSetsAreIndependentCopies()
        {
            var node = new Node<int, Event<int>>(new Event<int>(1));
            node.IncomingEdges.Add(10);

            var clone = (Node<int, Event<int>>)node.CloneObject();
            clone.IncomingEdges.Add(11);

            node.IncomingEdges.Count.ShouldBe(1);
            clone.IncomingEdges.Count.ShouldBe(2);
        }
    }
}

File: rust/primitives/tests/node_tests.rs
//! Ports of `.../Entities/NodeTests.cs`. The null-content constructor guard has
//! no Rust counterpart and is omitted; the C# `IncomingEdges`/`OutgoingEdges`
//! throwing accessors map to the `Result`-returning `incoming_edges` /
//! `outgoing_edges`.

use indexmap::IndexSet;
use zametek_maths_graphs_primitives::{Event, Node, NodeType};

type Nd = Node<i32, Event<i32>>;

#[test]
fn node_given_ctor_with_content_only_then_node_type_is_normal() {
    let node = Nd::new(Event::new(1));
    assert_eq!(node.node_type(), NodeType::Normal);
    assert_eq!(node.id(), 1);
}

#[test]
fn node_given_incoming_edges_when_start_node_then_err() {
    let node = Nd::with_type(NodeType::Start, Event::new(1));
    assert!(node.incoming_edges().is_err());
}

#[test]
fn node_given_incoming_edges_when_isolated_node_then_err() {
    let node = Nd::with_type(NodeType::Isolated, Event::new(1));
    assert!(node.incoming_edges().is_err());
}

#[test]
fn node_given_outgoing_edges_when_end_node_then_err() {
    let node = Nd::with_type(NodeType::End, Event::new(1));
    assert!(node.outgoing_edges().is_err());
}

#[test]
fn node_given_outgoing_edges_when_isolated_node_then_err() {
    let node = Nd::with_type(NodeType::Isolated, Event::new(1));
    assert!(node.outgoing_edges().is_err());
}

#[test]
fn node_given_set_node_type_then_node_type_changes() {
    let mut node = Nd::with_type(NodeType::Start, Event::new(1));
    node.set_node_type(NodeType::Normal);
    assert_eq!(node.node_type(), NodeType::Normal);
}

#[test]
fn node_given_equals_when_same_id_type_and_edges_then_equal() {
    let mut node1 = Nd::new(Event::new(1));
    node1.incoming.insert(10);
    node1.outgoing.insert(20);
    let mut node2 = Nd::new(Event::new(1));
    node2.incoming.insert(10);
    node2.outgoing.insert(20);
    assert_eq!(node1, node2);
}

#[test]
fn node_given_equals_when_different_node_type_then_not_equal() {
    let node1 = Nd::with_type(NodeType::Normal, Event::new(1));
    let node2 = Nd::with_type(NodeType::End, Event::new(1));
    assert_ne!(node1, node2);
}

#[test]
fn node_given_equals_when_different_edges_then_not_equal() {
    let mut node1 = Nd::new(Event::new(1));
    node1.incoming.insert(10);
    let mut node2 = Nd::new(Event::new(1));
    node2.incoming.insert(11);
    assert_ne!(node1, node2);
}

#[test]
fn node_given_clone_then_type_content_and_edges_preserved() {
    let mut node = Nd::with_type(NodeType::Normal, Event::with_times(1, Some(2), Some(3)));
    node.incoming.insert(10);
    node.outgoing.insert(20);

    let clone = node.clone();

    assert_eq!(clone.node_type(), NodeType::Normal);
    assert_eq!(clone.content.earliest_finish_time, Some(2));
    assert_eq!(clone.incoming, IndexSet::from([10]));
    assert_eq!(clone.outgoing, IndexSet::from([20]));
}

#[test]
fn node_given_clone_then_edge_sets_are_independent_copies() {
    let mut node = Nd::new(Event::new(1));
    node.incoming.insert(10);

    let mut clone = node.clone();
    clone.incoming.insert(11);

    assert_eq!(node.incoming.len(), 1);
    assert_eq!(clone.incoming.len(), 2);
}
